src/background/adblocker.js

```diff
@@ -54,31 +54,42 @@ function getEnabledEngines(config) {
   return [];
 }
 
-async function reloadMainEngine() {
-  const enabledEngines = getEnabledEngines(options);
-
-  if (enabledEngines.length) {
-    engines.replace(
-      engines.MAIN_ENGINE,
-      (
-        await Promise.all(
-          enabledEngines.map((id) =>
-            engines.init(id).catch(() => {
-              console.error(`[adblocker] failed to load engine: ${id}`);
-              return null;
-            }),
-          ),
-        )
-      ).filter((engine) => engine),
-    );
+function reloadMainEngine() {
+  clearTimeout(reloadMainEngine.timeout);
+
+  // Debounce the reloading of the main engine to avoid multiple reloads
+  // when the user changes multiple options in a short period of time
+  reloadMainEngine.timeout = setTimeout(
```

Comment:
Perhaps this could be a use case for requestIdleCallback (unfortunately not ready for Safari, but you can still fallback to setTimeout):

https://developer.mozilla.org/en-US/docs/Web/API/Window/requestIdleCallback

You can specify a timeout like the two seconds here, so you still have an upper bound. It will not batch multiple updates together, by default. So, it is not a drop-in replacement. But it could help to let the browser figure out an idle period to run the CPU bound computation.

Reply:
Service workers don't support it. Also, there is no problem with idle on the background side, it is on the web page context, which in Chrome is separated. 

The other thing is to denounce changes, as a user can click fast enough on option toggles, so main engine must be reloaded each time.